torch/testing/_internal/common_utils.py

```diff
@@ -691,12 +689,121 @@ def read_and_process():
+# types, so tensors with dtypes float16 and int64 produce tensors of
+# dtype float16 when added together.
+# If we used typical PyTorch type promotion to determine the dtype that tensors
+# should be compared in, however, then the test framework would assert
+# that tensors which are actually very different were "equal" to each other.
+# For example, torch.tensor((50000), dtype=torch.long) would be "equal" to
+# torch.tensor(49984., dtype=torch.float16), regardless of the rtol and atol.
+# These cases are admittedly, unlikely, but the test framework is not intended
+# to reflect only "likely" scenarios.
+# To preserve numerical accuracy, instead of following PyTorch's type
+# promotion rules this function follows NumPy's, which are intended to preserve
+# numerics. It special-cases bfloat16 since NumPy doesn't support bfloat16.
+# TODO: if we take a NumPy dependency then this function can be greatly
+# simplified.
+def get_comparison_dtype(a, b):
+    class _TypeMeta:
+        __slots__ = ['rank', 'kind', 'width']
+
+        def __init__(self, rank, kind, width):
+            self.rank = rank
+            self.kind = kind
+            self.width = width
+
+    # Maps torch dtypes to their rank, kind and width
+    type_meta = {
+        torch.complex128 : _TypeMeta(11, 'complex', 64),
+        torch.complex64  : _TypeMeta(10, 'complex', 32),
+        torch.float64    : _TypeMeta(9, 'float', 64),
+        torch.float32    : _TypeMeta(8, 'float', 32),
+        torch.bfloat16   : _TypeMeta(7, 'float', 16),
+        torch.float16    : _TypeMeta(6, 'float', 16),
+        torch.int64      : _TypeMeta(5, 'int', 64),
+        torch.int32      : _TypeMeta(4, 'int', 32),
+        torch.int16      : _TypeMeta(3, 'int', 16),
+        torch.int8       : _TypeMeta(2, 'int', 8),
+        torch.uint8      : _TypeMeta(1, 'uint', 8),
+        torch.bool       : _TypeMeta(0, 'uint', 1),
+    }
+
+    # Acquires common meta and acquires widest width
+    a_meta = type_meta[a.dtype]
+    b_meta = type_meta[b.dtype]
+    high_meta = a_meta if a_meta.rank >= b_meta.rank else b_meta
+    low_meta = b_meta if high_meta is a_meta else a_meta
+
+    # Short-circuits for boolean comparisons
+    if high_meta.rank == 0:
+        return torch.bool
+
+    # Acquires width
+    width = max(high_meta.width, low_meta.width)
+
+    # Adjusts width
+    # Doubles int width when upcasting
+    if ((high_meta.kind == 'float' or high_meta.kind == 'complex') and
+       low_meta.kind == 'int'):
+        width = min(64, max(low_meta.width * 2, high_meta.width))
+
+    # Doubles uint width when upcasting
+    if high_meta.kind != 'uint' and low_meta.kind == 'uint':
+        width = min(64, max(low_meta.width * 2, high_meta.width))
+
+    # Accounts for complex types being represented as double width
+    if high_meta.kind == 'complex':
+        width *= 2
+
+    # Accounts for bfloat16 x float16
+    if high_meta.rank == 7 and low_meta.rank == 6:
```

Comment:
this might be less brittle to additional dtypes if it checked for float and width==16

Reply:
Both bfloat16 and float16 have the same width, so we're trying to detect the exact bfloat16 x float16 case. This could be written as:

```
if (a.dtype is torch.bfloat16 or torch.float16) and (b.dtype is torch.bloat16 or torch.float16) and (a.dtype is not b.dtype):
```

Reply:
(mooted because function will be replaced with a table)